v2/types/types.go

```
@@ -364,6 +365,60 @@ type Type struct {
 
 	// The underlying Go type.
 	GoType gotypes.Type
+
+	// The reference to Multiverse
+	multiverse *multiverse
+}
+
+// multiverse holds Type definitions that were not found in the imported code but are needed during
+// generation. For example the imported code my have T but not *T, while the generated code needs *T. This
+// can't be part of the main Universe because that is a simple map, with no locking, and we all know what happens
+// when you modify a map while it is being iterated. Sharing this multiverse across Types ensures that there's
+// a single *Type for every type, and maintains the invariant that PointerTo(String) == PointerTo(String).
+type multiverse struct {
+	real      Universe
+	mu        sync.Mutex
+	synthetic map[string]*Type
+}
+
+// NeMultiverse creates a multiverse.
+func (t *Type) NewMultiverse(u Universe) {
+	if t == nil {
```

why not just let it panic - calling methods on nil is generally not allowed